addons/crm/static/src/js/forecast/forecast_model_extension.js

```diff
@@ -0,0 +1,116 @@
+/** @odoo-module */
+import ActionModel from 'web.ActionModel';
+
+/**
+ * This file contains the logic behind a special "Forecast" filter.
+ * Any such filter should set the context key {forecast_filter: 1}
+ * Another context key must also be set for the view using this model extension:
+ * {forecast_field: "date_field_name"}, which represents the date/time field on which
+ * the forecast should be applied
+ *
+ * The main purpose is to be able to modify the domain depending on the groupby granularity
+ * when the field is a `date/time` field and is the `forecast_field`. The domain should filter
+ * records from the period (granularity) containing "now", or those where the forecast_field
+ * is not set.
+ * example:
+ *  today:          2021-04-21
+ *  granularity:    month
+ *  field name:     date_field
+ *  -> the domain would be: ['&', ['date_field', '=', false], ['date_field', '>=', '2021-04-01']]
+ */
+const DATE_FORMAT = {
+    datetime: "YYYY-MM-DD HH:mm:ss",
+    date: "YYYY-MM-DD",
+};
+class ForecastModelExtension extends ActionModel.Extension {
+    /**
+     * @override
+     * @returns {any}
+     */
+    get(property) {
```

Comment:
In the end somehow if ``prepareState`` calls super, we could do it here (or remove both). Anyway no real effect, so your choice :)

Reply:
```python
get(property) {
        switch (property) {
            case "domain": return this.getDomain();
            default: return super.get(...arguments);
        }
    }
```
I did it like this (default of the switch case), because the way I see it, it is not the responsibility of `get` to merge multiples values of the same property, the latest child should overwrite every parents in case the same property is handled during the inheritance. (because if one wants unrelated domain bits, one should create multiple model_extensions, not extend an existing one)

The `search_panel_model_extension` recovers domain bits from every other model_extension and merge them here:
https://github.com/odoo/odoo/blob/75efaca796324707f18575dc7cce52b963e247df/addons/web/static/src/legacy/js/views/search_panel_model_extension.js#L607-L611

Reply:
I was more thing about "other" properties, not the same property to merge. Merging properties is out of scope indeed.

Reply:
And indeed it is correct, read it a bit too fast actually.